nemo/collections/nlp/modules/common/megatron/retrieval_service.py

```diff
@@ -13,64 +13,174 @@
 # limitations under the License.
 
 import abc
+import base64
+import json
+import logging
+import pickle
+import threading
 import time
 from typing import List, Union
 
 import faiss
 import numpy as np
+import requests
 import torch
+from flask import Flask, jsonify, request
+from flask_restful import Api, Resource
 from sentence_transformers import SentenceTransformer
 
 from nemo.collections.common.tokenizers.tokenizer_spec import TokenizerSpec
 from nemo.collections.nlp.data.language_modeling.megatron.indexed_retrieval_dataset import MMapRetrievalIndexedDataset
-from nemo.utils import logging
+
+log = logging.getLogger('werkzeug')
+log.setLevel(logging.ERROR)
+
+lock = threading.Lock()
+headers = {"Content-Type": "application/json"}
+
+PORT_NUM = 17179
+PORT_NUM_DYN = 17180
+PORT_NUM_BERT = 17181
+
+
+def request_data(data, port=PORT_NUM):
+    resp = requests.put('http://localhost:{}/knn'.format(port), data=json.dumps(data), headers=headers)
+    return resp.json()
 
 
 class RetrievalService:
     @abc.abstractmethod
-    def get_knn(self, query: Union[List[str], str, torch.Tensor]):
+    def get_knn(self, query: Union[List[str], str, torch.Tensor], neighbors: int):
         pass
 
+    @abc.abstractmethod
+    def add_docs_to_index(self, docs: List[str], add_eos: bool = True):
+        """
+        Add documents to the Faiss index
+        Args:
+            docs: List[str], list of documents that is going to be added to the index
+            add_eos: bool, whether add the eos in the end
+        """
+        raise NotImplementedError()
 
-class FaissRetrievalService(RetrievalService):
+
+class ChunkStore:
+    def __init__(self, chunk_size, pad_id):
+        self.store = {}
+        self._count = 0
+        self.no_retrieval = np.ones(2 * chunk_size, dtype=np.int64) * pad_id
```

Comment:
What is the use case of having this `self.no_retrieval`? I can see that it ensures it always has atleast one `dummy` element